<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper for processing coverage — coverage_process • dasper</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Wrapper for processing coverage — coverage_process" />
<meta property="og:description" content="coverage_process wraps all &quot;coverage_&quot; prefixed functions in
dasper. This is designed to simplify processing of the coverage data
for those familiar or uninterested with the intermediates." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dasper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/dasper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dzhang32/dasper/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper for processing coverage</h1>
    <small class="dont-index">Source: <a href='https://github.com/dzhang32/dasper/blob/master/R/coverage_process.R'><code>R/coverage_process.R</code></a></small>
    <div class="hidden name"><code>coverage_process.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>coverage_process</code> wraps all "coverage_" prefixed functions in
<a href='dasper.html'>dasper</a>. This is designed to simplify processing of the coverage data
for those familiar or uninterested with the intermediates.</p>
    </div>

    <pre class="usage"><span class='fu'>coverage_process</span><span class='op'>(</span>
  <span class='va'>junctions</span>,
  <span class='va'>ref</span>,
  unannot_width <span class='op'>=</span> <span class='fl'>20</span>,
  <span class='va'>coverage_paths_case</span>,
  <span class='va'>coverage_paths_control</span>,
  coverage_chr_control <span class='op'>=</span> <span class='cn'>NULL</span>,
  load_func <span class='op'>=</span> <span class='va'>.coverage_load</span>,
  bp_param <span class='op'>=</span> <span class='fu'>BiocParallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span><span class='op'>(</span><span class='op'>)</span>,
  norm_const <span class='op'>=</span> <span class='fl'>1</span>,
  score_func <span class='op'>=</span> <span class='va'>.zscore</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>junctions</th>
      <td><p>junction data as a
<a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object.</p></td>
    </tr>
    <tr>
      <th>ref</th>
      <td><p>either path to gtf/gff3 or object of class <a href='https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html'>TxDb-class</a>.</p></td>
    </tr>
    <tr>
      <th>unannot_width</th>
      <td><p>integer scalar determining the width of the region to
obtain coverage from when the end of of a junction does not overlap an
existing exon.</p></td>
    </tr>
    <tr>
      <th>coverage_paths_case</th>
      <td><p>paths to the BigWig files containing the
coverage of your case samples. Must be the same length and order to the
samples in <code>junctions</code>.</p></td>
    </tr>
    <tr>
      <th>coverage_paths_control</th>
      <td><p>paths to the BigWig files</p></td>
    </tr>
    <tr>
      <th>coverage_chr_control</th>
      <td><p>either "chr" or "no_chr", indicating the
chromosome format of control coverage data. Only used if you know the
chromosome format of the control BigWig files is different to that of
your junctions.</p></td>
    </tr>
    <tr>
      <th>load_func</th>
      <td><p>a function to use to load coverage. Currently only for
internal use to increase testing speed.</p></td>
    </tr>
    <tr>
      <th>bp_param</th>
      <td><p>a
<a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html'>BiocParallelParam-class</a> instance
denoting whether to parallelise the loading of coverage across BigWig files.</p></td>
    </tr>
    <tr>
      <th>norm_const</th>
      <td><p>numeric scaler to add to the normalisation coverage to avoid dividing
by 0s and resulting NaN or Inf values.</p></td>
    </tr>
    <tr>
      <th>score_func</th>
      <td><p>function to score junctions by their abnormality. By default,
will use a z-score but can be switched to a user-defined function. This
function must take as input an <code>x</code> and <code>y</code> argument, containing case and
control counts respectively. This must return a numeric vector equal to the
length of <code>x</code> with elements corresponding to a abnormality of each junction.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments passed to <code>score_func</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object containing junction data with coverage scores.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Coverage processing functions
<code><a href='coverage_norm.html'>coverage_norm</a>()</code>,
<code><a href='coverage_score.html'>coverage_score</a>()</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"ref"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='co'># use Genomic state to load txdb (GENCODE v31)</span>
    <span class='va'>ref</span> <span class='op'>&lt;-</span> <span class='fu'>GenomicState</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomicState/man/GenomicStateHub.html'>GenomicStateHub</a></span><span class='op'>(</span>version <span class='op'>=</span> <span class='st'>"31"</span>, genome <span class='op'>=</span> <span class='st'>"hg38"</span>, filetype <span class='op'>=</span> <span class='st'>"TxDb"</span><span class='op'>)</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
    <span class='co'># convert seqlevels to match junctions</span>
    <span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>ref</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>stringr</span><span class='fu'>::</span><span class='fu'><a href='https://stringr.tidyverse.org/reference/str_replace.html'>str_replace</a></span><span class='op'>(</span><span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>ref</span><span class='op'>)</span>, <span class='st'>"chr"</span>, <span class='st'>""</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2020-09-03</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"junctions_processed"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>junctions_processed</span> <span class='op'>&lt;-</span>
        <span class='fu'><a href='junction_process.html'>junction_process</a></span><span class='op'>(</span>
            <span class='va'>junctions_example</span>,
            <span class='va'>ref</span>,
            count_thresh <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"raw"</span> <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>,
            n_samp <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"raw"</span> <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>,
            width_range <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>25</span>, <span class='fl'>1000000</span><span class='op'>)</span>,
            types <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ambig_gene"</span>, <span class='st'>"unannotated"</span><span class='op'>)</span>,
        <span class='op'>)</span>
    <span class='fu'>GenomeInfoDb</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqlevels</a></span><span class='op'>(</span><span class='va'>junctions_processed</span><span class='op'>)</span> <span class='op'>&lt;-</span>
        <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"chr"</span>, <span class='fu'>GenomeInfoDb</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqlevels</a></span><span class='op'>(</span><span class='va'>junctions_processed</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-10-14 17:54:22 - Filtering junctions..."
#&gt; [1] "2020-10-14 17:54:22 - by count..."
#&gt; [1] "2020-10-14 17:54:23 - by width..."
#&gt; [1] "2020-10-14 17:54:23 - done!"
#&gt; [1] "# Annotating junctions ----------------------------------------------------"
#&gt; [1] "2020-10-14 17:54:23 - Obtaining co-ordinates of annotated exons and junctions from gtf/gff3..."
#&gt; [1] "2020-10-14 17:54:41 - Getting junction annotation using overlapping exons..."
#&gt; [1] "2020-10-14 17:54:43 - Tidying junction annotation..."
#&gt; [1] "2020-10-14 17:54:43 - Deriving junction categories..."
#&gt; [1] "2020-10-14 17:54:44 - done!"
#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-10-14 17:54:44 - Filtering junctions..."
#&gt; [1] "2020-10-14 17:54:44 - by type..."
#&gt; [1] "2020-10-14 17:54:44 - done!"
#&gt; [1] "# Normalise junctions -----------------------------------------------------"
#&gt; [1] "2020-10-14 17:54:44 - Clustering junctions..."
#&gt; [1] "2020-10-14 17:54:44 - Normalising junction counts..."
#&gt; [1] "2020-10-14 17:54:44 - done!"
#&gt; [1] "# Score junctions ---------------------------------------------------------"
#&gt; [1] "2020-10-14 17:54:44 - Calculating the direction of change of junctions..."
#&gt; [1] "2020-10-14 17:54:44 - Generating junction abnormality score..."
#&gt; [1] "2020-10-14 17:54:44 - done!"</div><div class='input'>
<span class='co'># obtain path to example bw on recount2</span>
<span class='va'>url</span> <span class='op'>&lt;-</span> <span class='fu'>recount</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/recount/man/download_study.html'>download_study</a></span><span class='op'>(</span>
    project <span class='op'>=</span> <span class='st'>"SRP012682"</span>,
    type <span class='op'>=</span> <span class='st'>"samples"</span>,
    download <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span>

<span class='va'>bw_path</span> <span class='op'>&lt;-</span> <span class='fu'>dasper</span><span class='fu'>:::</span><span class='fu'>.file_cache</span><span class='op'>(</span><span class='va'>url</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>)</span>

<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"junctions_w_coverage"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>junctions_w_coverage</span> <span class='op'>&lt;-</span>
        <span class='fu'>coverage_process</span><span class='op'>(</span>
            <span class='va'>junctions_processed</span>,
            <span class='va'>ref</span>,
            coverage_paths_case <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='va'>bw_path</span>, <span class='fl'>2</span><span class='op'>)</span>,
            coverage_paths_control <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='va'>bw_path</span>, <span class='fl'>3</span><span class='op'>)</span>
        <span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; [1] "# Loading and normalising coverage ---------------------------------------------"
#&gt; [1] "2020-10-14 17:54:45 - Obtaining exonic and intronic regions to load coverage from..."
#&gt; [1] "2020-10-14 17:54:45 - Obtaining regions to use to normalise coverage..."</div><div class='output co'>#&gt; <span class='warning'>Warning: The 2 combined objects have no sequence levels in common. (Use</span>
#&gt; <span class='warning'>  suppressWarnings() to suppress this warning.)</span></div><div class='output co'>#&gt; [1] "2020-10-14 17:54:50 - Loading coverage..."</div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; [1] "2020-10-14 17:54:53 - Normalising coverage..."
#&gt; [1] "2020-10-14 17:54:53 - done!"
#&gt; [1] "# Scoring coverage ---------------------------------------------"
#&gt; [1] "2020-10-14 17:54:53 - Generating coverage abnormality score..."
#&gt; [1] "2020-10-14 17:54:53 - Obtaining regions with greatest coverage dysruption..."
#&gt; [1] "2020-10-14 17:54:54 - done!"</div><div class='input'>
<span class='va'>junctions_w_coverage</span>
</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 4705 2 
#&gt; metadata(0):
#&gt; assays(6): raw norm ... coverage_region coverage_score
#&gt; rownames: NULL
#&gt; rowData names(16): in_ref gene_id_start ... index clusters
#&gt; colnames(2): count_1 count_2
#&gt; colData names(2): samp_id case_control</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


