<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>For each junction, obtains the regions to obtain coverage — junction_cov_norm • dasper</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="For each junction, obtains the regions to obtain coverage — junction_cov_norm" />
<meta property="og:description" content="junction_cov_norm obtains regions of interest for each junction where
coverage disruptions would be expected. If ends of junctions are annotated,
it will use the overlapping exon definitions, picking the shortest exon when
multiple overlap one end. If unannotated, will use a user-defined width
(i.e. 20bp). To compare between samples (case/controls) coverage is
normalised to a set region. By default, the boundaries of each gene
associated to a junction are used as the region to normalise to." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dasper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/dasper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dzhang32/dasper/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>For each junction, obtains the regions to obtain coverage</h1>
    <small class="dont-index">Source: <a href='https://github.com/dzhang32/dasper/blob/master/R/junction_cov_norm.R'><code>R/junction_cov_norm.R</code></a></small>
    <div class="hidden name"><code>junction_cov_norm.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>junction_cov_norm</code> obtains regions of interest for each junction where
coverage disruptions would be expected. If ends of junctions are annotated,
it will use the overlapping exon definitions, picking the shortest exon when
multiple overlap one end. If unannotated, will use a user-defined width
(i.e. 20bp). To compare between samples (case/controls) coverage is
normalised to a set region. By default, the boundaries of each gene
associated to a junction are used as the region to normalise to.</p>
    </div>

    <pre class="usage"><span class='fu'>junction_cov_norm</span>(
  <span class='kw'>junctions</span>,
  <span class='kw'>ref</span>,
  unannot_width = <span class='fl'>20</span>,
  <span class='kw'>cov_paths_case</span>,
  <span class='kw'>cov_paths_control</span>,
  cov_chr_control = <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>junctions</th>
      <td><p>junction data as a
<a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object.</p></td>
    </tr>
    <tr>
      <th>ref</th>
      <td><p>either path to gtf/gff3 or object of class <code>TxDb</code> imported using
<code><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></code>.</p></td>
    </tr>
    <tr>
      <th>unannot_width</th>
      <td><p>integer scalar determining the width of the region to
obtain coverage from when the end of of a junction does not overlap an
existing exon.</p></td>
    </tr>
    <tr>
      <th>cov_paths_case</th>
      <td><p>paths to the BigWig/BAM files containing the coverage of
your case samples. Must be the same length and order to the rows of the
assays in <code>junctions</code>.</p></td>
    </tr>
    <tr>
      <th>cov_paths_control</th>
      <td><p>paths to the BigWig/BAM files</p></td>
    </tr>
    <tr>
      <th>cov_chr_control</th>
      <td><p>either "chr" or "no_chr", indicating the chromosome
format of control coverage data. To be used if you know the chromosome
format of the control BigWig/BAM files is different to that of your
junctions.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list containing sublists, one for cases and the other controls. Each
sublist contains 3 matrices, corresponding the coverage for each sample
across the 2 exons and intron associated with every junction.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'># leave this as not run for now to save time for R CMD check</span>
<span class='kw'>ref</span> <span class='op'>&lt;-</span> <span class='st'>"ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.gtf.gz"</span>
<span class='kw'>ref</span> <span class='op'>&lt;-</span> <span class='kw'>GenomicFeatures</span>::<span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></span>(<span class='kw'>ref</span>)
<span class='kw'>junctions</span> <span class='op'>&lt;-</span> <span class='fu'>cov_norm</span>(<span class='kw'>junctions_annot_example</span>, <span class='kw'>ref</span>)
<span class='kw'>junctions</span>
}
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


