<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper function for processing junctions — junction_process • dasper</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Wrapper function for processing junctions — junction_process" />
<meta property="og:description" content="junction_process wraps all &quot;junction_&quot; prefixed functions in
dasper except junction_load. This is designed to simplify
processing of the junction data for those familiar or uninterested with the
intermediates." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dasper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/dasper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dzhang32/dasper/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper function for processing junctions</h1>
    <small class="dont-index">Source: <a href='https://github.com/dzhang32/dasper/blob/master/R/junction_process.R'><code>R/junction_process.R</code></a></small>
    <div class="hidden name"><code>junction_process.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>junction_process</code> wraps all "junction_" prefixed functions in
<code>dasper</code> except <a href='junction_load.html'>junction_load</a>. This is designed to simplify
processing of the junction data for those familiar or uninterested with the
intermediates.</p>
    </div>

    <pre class="usage"><span class='fu'>junction_process</span>(
  <span class='kw'>junctions</span>,
  <span class='kw'>ref</span>,
  count_thresh = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(raw = <span class='fl'>5</span>),
  n_samp = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(raw = <span class='fl'>1</span>),
  width_range = <span class='kw'>NULL</span>,
  types = <span class='kw'>NULL</span>,
  regions = <span class='kw'>NULL</span>,
  score_func = <span class='kw'>.zscore</span>,
  <span class='kw'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>junctions</th>
      <td><p>junction data as a
<a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object.</p></td>
    </tr>
    <tr>
      <th>ref</th>
      <td><p>either path to gtf/gff3 or object of class <code>TxDb</code> imported using
<code><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></code>.</p></td>
    </tr>
    <tr>
      <th>count_thresh</th>
      <td><p>named vector with names matching the names of the
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>assays</a></code> in <code>junctions</code>. Values
denote the number of counts below which a junction will be filtered out.</p></td>
    </tr>
    <tr>
      <th>n_samp</th>
      <td><p>named vector with names matching the names of the
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>assays</a></code> in <code>junctions</code>. Values
denotes number of samples that have to express the junction above the
<code>count_thresh</code> in order for that junction to not be filtered.</p></td>
    </tr>
    <tr>
      <th>width_range</th>
      <td><p>numeric vector of length 2. The first element denoting the
lower limit of junction width and the second the upper limit. Junctions
with widths outside this range will be filtered out.</p></td>
    </tr>
    <tr>
      <th>types</th>
      <td><p>any junctions matching these types, derived form
<code><a href='junction_annot.html'>junction_annot</a></code> e.g. "unannotated" will be filtered out.</p></td>
    </tr>
    <tr>
      <th>regions</th>
      <td><p>any junctions overlapping this set of regions (in a
<code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> format) will be filtered out.</p></td>
    </tr>
    <tr>
      <th>score_func</th>
      <td><p>function to score junctions by abnromality. By default,
will use a z-score. But this can be switched dependent on the format of the
user's junction data. Function must take as input an x and y argument. x
being the case junctions and y being the controls. This must return a
numeric vector equal to the length of x with each element corresponding to
a abnormality of a junction in each case sample.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments passed to <code>score_func</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object containing filtered, annotated, normalised junction data with scores.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'>ref</span> <span class='op'>&lt;-</span>
    <span class='st'>"ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.gtf.gz"</span>
<span class='kw'>junctions_processed</span> <span class='op'>&lt;-</span>
    <span class='fu'>junction_process</span>(
        <span class='kw'>junctions_example</span>,
        <span class='kw'>ref</span>,
        count_thresh = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"raw"</span> = <span class='fl'>5</span>),
        n_samp = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"raw"</span> = <span class='fl'>1</span>),
        width_range = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>25</span>, <span class='fl'>1000000</span>),
        types = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ambig_gene"</span>, <span class='st'>"unannotated"</span>),
    )</div><div class='output co'>#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-08-06 10:24:43 - Filtering junctions..."
#&gt; [1] "2020-08-06 10:24:43 - by count..."
#&gt; [1] "2020-08-06 10:24:43 - by width..."
#&gt; [1] "2020-08-06 10:24:43 - done!"
#&gt; [1] "# Annotating junctions ----------------------------------------------------"
#&gt; [1] "2020-08-06 10:24:43 - Obtaining co-ordinates of annotated exons and junctions from gtf/gff3..."
#&gt; [1] "2020-08-06 10:24:43 - Importing gtf/gff3 as a TxDb..."</div><div class='output co'>#&gt; <span class='message'>Import genomic features from the file as a GRanges object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Prepare the 'metadata' data frame ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Make the TxDb object ... </span></div><div class='output co'>#&gt; <span class='warning'>Warning: The "phase" metadata column contains non-NA values for features of type</span>
#&gt; <span class='warning'>  stop_codon. This information was ignored.</span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; [1] "2020-08-06 10:26:00 - Getting junction annotation using overlapping exons..."
#&gt; [1] "2020-08-06 10:26:00 - Tidying junction annotation..."
#&gt; [1] "2020-08-06 10:26:01 - Deriving junction categories..."
#&gt; [1] "2020-08-06 10:26:01 - done!"
#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-08-06 10:26:01 - Filtering junctions..."
#&gt; [1] "2020-08-06 10:26:01 - by type..."
#&gt; [1] "2020-08-06 10:26:01 - done!"
#&gt; [1] "# Normalise junctions -----------------------------------------------------"
#&gt; [1] "2020-08-06 10:26:01 - Clustering junctions..."
#&gt; [1] "2020-08-06 10:26:01 - Normalising junction counts..."
#&gt; [1] "2020-08-06 10:26:01 - done!"
#&gt; [1] "# Score junctions ---------------------------------------------------------"
#&gt; [1] "2020-08-06 10:26:01 - Generating junction abnormality score..."
#&gt; [1] "2020-08-06 10:26:01 - done!"</div><div class='input'><span class='kw'>junctions_processed</span></div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 4706 2 
#&gt; metadata(0):
#&gt; assays(3): raw norm score
#&gt; rownames: NULL
#&gt; rowData names(16): in_ref gene_id_start ... index clusters
#&gt; colnames(2): count_1 count_2
#&gt; colData names(2): samp_id case_control</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


