<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Detecting outlier junctions — outlier_detect • dasper</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Detecting outlier junctions — outlier_detect" />
<meta property="og:description" content="outlier_detect will use the features in
assays named
feature_names as input into an unsupervised outlier detection algorithm to
score each junction based on how outlier-y it looks in relation to other
junctions in the patient. The default expected score and coverage_score
features can be calculated using the junction_process and
coverage_process respectively." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dasper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/dasper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dzhang32/dasper/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detecting outlier junctions</h1>
    <small class="dont-index">Source: <a href='https://github.com/dzhang32/dasper/blob/master/R/outlier_detect.R'><code>R/outlier_detect.R</code></a></small>
    <div class="hidden name"><code>outlier_detect.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>outlier_detect</code> will use the features in
<a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>assays</a> named
<code>feature_names</code> as input into an unsupervised outlier detection algorithm to
score each junction based on how outlier-y it looks in relation to other
junctions in the patient. The default expected <code>score</code> and <code>coverage_score</code>
features can be calculated using the <a href='junction_process.html'>junction_process</a> and
<a href='coverage_process.html'>coverage_process</a> respectively.</p>
    </div>

    <pre class="usage"><span class='fu'>outlier_detect</span><span class='op'>(</span>
  <span class='va'>junctions</span>,
  feature_names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"score"</span>, <span class='st'>"coverage_score"</span><span class='op'>)</span>,
  bp_param <span class='op'>=</span> <span class='fu'>BiocParallel</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>junctions</th>
      <td><p>junction data as a
<a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object.</p></td>
    </tr>
    <tr>
      <th>feature_names</th>
      <td><p>names of assays in <code>junctions</code> that are to be used as
input into the outlier detection model.</p></td>
    </tr>
    <tr>
      <th>bp_param</th>
      <td><p>a
<a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html'>BiocParallelParam-class</a> instance
denoting whether to parallelise the calculating of outlier scores across
samples.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional arguments passed to the outlier detection model
(isolation forest) for setting parameters.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>junctions as a
<a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object with additional assays named <code>outlier_score</code>. The lower the outlier
score, the more of an outlier a junction is observed to be.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>for more details on the isolation forest model used:
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.IsolationForest.html</p>
<p>Outlier processing functions
<code><a href='outlier_aggregate.html'>outlier_aggregate</a>()</code>,
<code><a href='outlier_process.html'>outlier_process</a>()</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='va'>.Platform</span><span class='op'>$</span><span class='va'>OS.type</span> <span class='op'>!=</span> <span class='st'>"windows"</span><span class='op'>)</span> <span class='op'>{</span>

    <span class='co'># tell reticulate to use the python3 install</span>
    <span class='co'># if windows skip this step</span>
    <span class='fu'>reticulate</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/reticulate/man/use_python.html'>use_python</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/Sys.which.html'>Sys.which</a></span><span class='op'>(</span><span class='st'>"python3"</span><span class='op'>)</span>, required <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='op'>}</span>

<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"ref"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>ref</span> <span class='op'>&lt;-</span> <span class='st'>"ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.gtf.gz"</span>
    <span class='va'>ref</span> <span class='op'>&lt;-</span> <span class='fu'>GenomicFeatures</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>makeTxDbFromGFF</a></span><span class='op'>(</span><span class='va'>ref</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; <span class='message'>Import genomic features from the file as a GRanges object ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Prepare the 'metadata' data frame ... </span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='output co'>#&gt; <span class='message'>Make the TxDb object ... </span></div><div class='output co'>#&gt; <span class='warning'>Warning: The "phase" metadata column contains non-NA values for features of type</span>
#&gt; <span class='warning'>  stop_codon. This information was ignored.</span></div><div class='output co'>#&gt; <span class='message'>OK</span></div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"junctions_processed"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>junctions_processed</span> <span class='op'>&lt;-</span>
        <span class='fu'><a href='junction_process.html'>junction_process</a></span><span class='op'>(</span>
            <span class='va'>junctions_example</span>,
            <span class='va'>ref</span>,
            count_thresh <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"raw"</span> <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>,
            n_samp <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"raw"</span> <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>,
            width_range <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>25</span>, <span class='fl'>1000000</span><span class='op'>)</span>,
            types <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ambig_gene"</span>, <span class='st'>"unannotated"</span><span class='op'>)</span>,
        <span class='op'>)</span>
    <span class='fu'>GenomeInfoDb</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqlevels</a></span><span class='op'>(</span><span class='va'>junctions_processed</span><span class='op'>)</span> <span class='op'>&lt;-</span>
        <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"chr"</span>, <span class='fu'>GenomeInfoDb</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqlevels</a></span><span class='op'>(</span><span class='va'>junctions_processed</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-10-02 23:00:11 - Filtering junctions..."
#&gt; [1] "2020-10-02 23:00:11 - by count..."
#&gt; [1] "2020-10-02 23:00:11 - by width..."
#&gt; [1] "2020-10-02 23:00:11 - done!"
#&gt; [1] "# Annotating junctions ----------------------------------------------------"
#&gt; [1] "2020-10-02 23:00:11 - Obtaining co-ordinates of annotated exons and junctions from gtf/gff3..."
#&gt; [1] "2020-10-02 23:00:26 - Getting junction annotation using overlapping exons..."
#&gt; [1] "2020-10-02 23:00:26 - Tidying junction annotation..."
#&gt; [1] "2020-10-02 23:00:26 - Deriving junction categories..."
#&gt; [1] "2020-10-02 23:00:26 - done!"
#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-10-02 23:00:26 - Filtering junctions..."
#&gt; [1] "2020-10-02 23:00:26 - by type..."
#&gt; [1] "2020-10-02 23:00:26 - done!"
#&gt; [1] "# Normalise junctions -----------------------------------------------------"
#&gt; [1] "2020-10-02 23:00:26 - Clustering junctions..."
#&gt; [1] "2020-10-02 23:00:27 - Normalising junction counts..."
#&gt; [1] "2020-10-02 23:00:27 - done!"
#&gt; [1] "# Score junctions ---------------------------------------------------------"
#&gt; [1] "2020-10-02 23:00:27 - Calculating the direction of change of junctions..."
#&gt; [1] "2020-10-02 23:00:27 - Generating junction abnormality score..."
#&gt; [1] "2020-10-02 23:00:27 - done!"</div><div class='input'>
<span class='va'>url</span> <span class='op'>&lt;-</span> <span class='fu'>recount</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/recount/man/download_study.html'>download_study</a></span><span class='op'>(</span>
    project <span class='op'>=</span> <span class='st'>"SRP012682"</span>,
    type <span class='op'>=</span> <span class='st'>"samples"</span>,
    download <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span>
<span class='va'>bw_path</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span><span class='op'>(</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/basename.html'>basename</a></span><span class='op'>(</span><span class='va'>url</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span>

<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/files.html'>file.exists</a></span><span class='op'>(</span><span class='va'>bw_path</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='fu'><a href='https://rdrr.io/r/utils/download.file.html'>download.file</a></span><span class='op'>(</span><span class='va'>url</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>, <span class='va'>bw_path</span><span class='op'>)</span>
<span class='op'>}</span>

<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"junctions_w_coverage"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>junctions_w_coverage</span> <span class='op'>&lt;-</span>
        <span class='fu'><a href='coverage_process.html'>coverage_process</a></span><span class='op'>(</span>
            <span class='va'>junctions_processed</span>,
            <span class='va'>ref</span>,
            coverage_paths_case <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='va'>bw_path</span>, <span class='fl'>2</span><span class='op'>)</span>,
            coverage_paths_control <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='va'>bw_path</span>, <span class='fl'>3</span><span class='op'>)</span>
        <span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; [1] "# Loading and normalising coverage ---------------------------------------------"
#&gt; [1] "2020-10-02 23:00:27 - Obtaining exonic and intronic regions to load coverage from..."
#&gt; [1] "2020-10-02 23:00:27 - Obtaining regions to use to normalise coverage..."</div><div class='output co'>#&gt; <span class='warning'>Warning: The 2 combined objects have no sequence levels in common. (Use</span>
#&gt; <span class='warning'>  suppressWarnings() to suppress this warning.)</span></div><div class='output co'>#&gt; [1] "2020-10-02 23:00:32 - Loading coverage..."</div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Total AUC across all regions was 0. Make sure chromsome format matches between input regions and bigWig/BAM file.</span></div><div class='output co'>#&gt; [1] "2020-10-02 23:00:34 - Normalising coverage..."
#&gt; [1] "2020-10-02 23:00:34 - done!"
#&gt; [1] "# Scoring coverage ---------------------------------------------"
#&gt; [1] "2020-10-02 23:00:34 - Generating coverage abnormality score..."
#&gt; [1] "2020-10-02 23:00:35 - Obtaining regions with greatest coverage dysruption..."
#&gt; [1] "2020-10-02 23:00:35 - done!"</div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"junctions_w_outliers"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>junctions_w_outliers</span> <span class='op'>&lt;-</span> <span class='fu'>outlier_detect</span><span class='op'>(</span><span class='va'>junctions_w_coverage</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; [1] "2020-10-02 23:00:35 - generating outlier scores for sample 1/2"
#&gt; [1] "2020-10-02 23:00:35 - fitting outlier detection model with parameters: behaviour=deprecated, bootstrap=FALSE, contamination=auto, max_features=1, max_samples=auto, n_estimators=100, n_jobs=NULL, random_state=NULL, verbose=0, warm_start=FALSE"
#&gt; [1] "2020-10-02 23:00:35 - fitting outlier detection model with parameters: behaviour=deprecated, bootstrap=FALSE, contamination=auto, max_features=1, max_samples=auto, n_estimators=100, n_jobs=NULL, random_state=NULL, verbose=0, warm_start=FALSE"
#&gt; [1] "2020-10-02 23:00:35 - generating outlier scores for sample 2/2"
#&gt; [1] "2020-10-02 23:00:35 - fitting outlier detection model with parameters: behaviour=deprecated, bootstrap=FALSE, contamination=auto, max_features=1, max_samples=auto, n_estimators=100, n_jobs=NULL, random_state=NULL, verbose=0, warm_start=FALSE"
#&gt; [1] "2020-10-02 23:00:36 - fitting outlier detection model with parameters: behaviour=deprecated, bootstrap=FALSE, contamination=auto, max_features=1, max_samples=auto, n_estimators=100, n_jobs=NULL, random_state=NULL, verbose=0, warm_start=FALSE"
#&gt; [1] "2020-10-02 23:00:36 - done!"</div><div class='input'>
<span class='va'>junctions_w_outliers</span>
</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 4706 2 
#&gt; metadata(0):
#&gt; assays(7): raw norm ... coverage_score outlier_score
#&gt; rownames: NULL
#&gt; rowData names(16): in_ref gene_id_start ... index clusters
#&gt; colnames(2): count_1 count_2
#&gt; colData names(2): samp_id case_control</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


