<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Visualise RNA-seq data in a the form of a sashimi plot — plot_sashimi • dasper</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Visualise RNA-seq data in a the form of a sashimi plot — plot_sashimi" />
<meta property="og:description" content="plot_sashimi plots the splicing events and coverage over a specific
genes/transcript and/or genomic region of interest. The plotting are built
from ggplot2 functions." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dasper</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/dasper.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/dzhang32/dasper/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Visualise RNA-seq data in a the form of a sashimi plot</h1>
    <small class="dont-index">Source: <a href='https://github.com/dzhang32/dasper/blob/master/R/plot_sashimi.R'><code>R/plot_sashimi.R</code></a></small>
    <div class="hidden name"><code>plot_sashimi.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>plot_sashimi</code> plots the splicing events and coverage over a specific
genes/transcript and/or genomic region of interest. The plotting are built
from <code>ggplot2</code> functions.</p>
    </div>

    <pre class="usage"><span class='fu'>plot_sashimi</span><span class='op'>(</span>
  <span class='va'>junctions</span>,
  <span class='va'>ref</span>,
  <span class='va'>gene_tx_id</span>,
  case_id <span class='op'>=</span> <span class='cn'>NULL</span>,
  control_agg_func <span class='op'>=</span> <span class='va'>mean</span>,
  region <span class='op'>=</span> <span class='cn'>NULL</span>,
  annot_colour <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'>ggpubr</span><span class='fu'>::</span><span class='fu'><a href='https://rpkgs.datanovia.com/ggpubr/reference/get_palette.html'>get_palette</a></span><span class='op'>(</span><span class='st'>"jco"</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'>ggpubr</span><span class='fu'>::</span><span class='fu'><a href='https://rpkgs.datanovia.com/ggpubr/reference/get_palette.html'>get_palette</a></span><span class='op'>(</span><span class='st'>"npg"</span>, <span class='fl'>7</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>3</span>,
    <span class='fl'>2</span>, <span class='fl'>5</span>, <span class='fl'>6</span><span class='op'>)</span><span class='op'>]</span>, <span class='fu'>ggpubr</span><span class='fu'>::</span><span class='fu'><a href='https://rpkgs.datanovia.com/ggpubr/reference/get_palette.html'>get_palette</a></span><span class='op'>(</span><span class='st'>"jco"</span>, <span class='fl'>6</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>3</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>,
  digits <span class='op'>=</span> <span class='fl'>2</span>,
  count_label <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>junctions</th>
      <td><p>junction data as a
<a href='https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html'>RangedSummarizedExperiment-class</a>
object.</p></td>
    </tr>
    <tr>
      <th>ref</th>
      <td><p>either path to gtf/gff3 or object of class <a href='https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html'>TxDb-class</a>.</p></td>
    </tr>
    <tr>
      <th>gene_tx_id</th>
      <td><p>gene name as ensembl id, ensembl transcript id or gene
name.</p></td>
    </tr>
    <tr>
      <th>case_id</th>
      <td><p>list of one element. This must be a character vector
containing the</p></td>
    </tr>
    <tr>
      <th>control_agg_func</th>
      <td><p>list of one element. This must be a character vector
containing the</p></td>
    </tr>
    <tr>
      <th>region</th>
      <td><p>a <a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GenomicRanges</a> of length 1
that is used to filter the exons/junctions. Only those that overlap this
region are plotted.</p></td>
    </tr>
    <tr>
      <th>annot_colour</th>
      <td><p>character vector of colours for junction types. One value
must be supplied labelling each of the 7 possible types.</p></td>
    </tr>
    <tr>
      <th>digits</th>
      <td><p>used in <code>round</code>, specifying the number of digits to round the
junction counts to for visualisation purposes.</p></td>
    </tr>
    <tr>
      <th>count_label</th>
      <td><p>logical value specifying whether to add label the count of
each junction.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>ggplot</code> displaying the splicing (and coverage) surrounding the
transcript/region of interest.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"ref"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='co'># use Genomic state to load txdb (GENCODE v31)</span>
    <span class='va'>ref</span> <span class='op'>&lt;-</span> <span class='fu'>GenomicState</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomicState/man/GenomicStateHub.html'>GenomicStateHub</a></span><span class='op'>(</span>version <span class='op'>=</span> <span class='st'>"31"</span>, genome <span class='op'>=</span> <span class='st'>"hg38"</span>, filetype <span class='op'>=</span> <span class='st'>"TxDb"</span><span class='op'>)</span><span class='op'>[[</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>]</span>
    <span class='co'># convert seqlevels to match junctions</span>
    <span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>ref</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'>stringr</span><span class='fu'>::</span><span class='fu'><a href='https://stringr.tidyverse.org/reference/str_replace.html'>str_replace</a></span><span class='op'>(</span><span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>ref</span><span class='op'>)</span>, <span class='st'>"chr"</span>, <span class='st'>""</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; <span class='message'>snapshotDate(): 2020-09-03</span></div><div class='output co'>#&gt; <span class='message'>loading from cache</span></div><div class='input'>
<span class='kw'>if</span> <span class='op'>(</span><span class='op'>!</span><span class='fu'><a href='https://rdrr.io/r/base/exists.html'>exists</a></span><span class='op'>(</span><span class='st'>"junctions_processed"</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>
    <span class='va'>junctions_processed</span> <span class='op'>&lt;-</span>
        <span class='fu'><a href='junction_process.html'>junction_process</a></span><span class='op'>(</span>
            <span class='va'>junctions_example</span>,
            <span class='va'>ref</span>,
            count_thresh <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"raw"</span> <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>,
            n_samp <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"raw"</span> <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span>,
            width_range <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>25</span>, <span class='fl'>1000000</span><span class='op'>)</span>,
            types <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ambig_gene"</span>, <span class='st'>"unannotated"</span><span class='op'>)</span>,
        <span class='op'>)</span>
    <span class='fu'>GenomeInfoDb</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqlevels</a></span><span class='op'>(</span><span class='va'>junctions_processed</span><span class='op'>)</span> <span class='op'>&lt;-</span>
        <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"chr"</span>, <span class='fu'>GenomeInfoDb</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqlevels</a></span><span class='op'>(</span><span class='va'>junctions_processed</span><span class='op'>)</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-10-14 17:58:36 - Filtering junctions..."
#&gt; [1] "2020-10-14 17:58:36 - by count..."
#&gt; [1] "2020-10-14 17:58:36 - by width..."
#&gt; [1] "2020-10-14 17:58:36 - done!"
#&gt; [1] "# Annotating junctions ----------------------------------------------------"
#&gt; [1] "2020-10-14 17:58:36 - Obtaining co-ordinates of annotated exons and junctions from gtf/gff3..."
#&gt; [1] "2020-10-14 17:58:52 - Getting junction annotation using overlapping exons..."
#&gt; [1] "2020-10-14 17:58:53 - Tidying junction annotation..."
#&gt; [1] "2020-10-14 17:58:53 - Deriving junction categories..."
#&gt; [1] "2020-10-14 17:58:54 - done!"
#&gt; [1] "# Filtering junctions -----------------------------------------------------"
#&gt; [1] "2020-10-14 17:58:54 - Filtering junctions..."
#&gt; [1] "2020-10-14 17:58:54 - by type..."
#&gt; [1] "2020-10-14 17:58:54 - done!"
#&gt; [1] "# Normalise junctions -----------------------------------------------------"
#&gt; [1] "2020-10-14 17:58:54 - Clustering junctions..."
#&gt; [1] "2020-10-14 17:58:54 - Normalising junction counts..."
#&gt; [1] "2020-10-14 17:58:54 - done!"
#&gt; [1] "# Score junctions ---------------------------------------------------------"
#&gt; [1] "2020-10-14 17:58:54 - Calculating the direction of change of junctions..."
#&gt; [1] "2020-10-14 17:58:54 - Generating junction abnormality score..."
#&gt; [1] "2020-10-14 17:58:54 - done!"</div><div class='input'>
<span class='va'>sashimi_plot</span> <span class='op'>&lt;-</span> <span class='fu'>plot_sashimi</span><span class='op'>(</span>
    junctions <span class='op'>=</span> <span class='va'>junctions_processed</span>,
    ref <span class='op'>=</span> <span class='va'>ref</span>,
    gene_tx_id <span class='op'>=</span> <span class='st'>"ENSG00000241973.10"</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 108 row(s) containing missing values (geom_path).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 4 rows containing missing values (geom_label_repel).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 108 row(s) containing missing values (geom_path).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 4 rows containing missing values (geom_label_repel).</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


