#' Score coverage disruptions
#'
#' \code{coverage_score} will score disruptions in the coverage across the
#' intronic/exonic regions associated with each junction. This abnormality score
#' is generated by \code{score_func} and operates by comparing coverage
#' originating from patients to a set of controls. Then, for each junction it
#' obtains the score of the region with the highest absolute score/greatest
#' disruption.
#'
#' @inheritParams junction_annot
#' @inheritParams junction_score
#'
#' @param coverage list containing normalised coverage data that is outputted from
#'   \link{coverage_norm}.
#'
#' @return junctions as a
#'   \code{\link[SummarizedExperiment]{SummarizedExperiment}} object with
#'   additional assays named \code{coverage_region} and \code{coverage_score}.
#'   \code{coverage_region} labels the region of greatest disruption (1 = exon_start,
#'   2 = exon_end, 3 = intron) and \code{coverage_score} contains the scores of the
#'   region with the greatest disruption.
#'
#' @examples
#'
#' \dontrun{
#' # leave this as not run for now to save time for R CMD check
#' ref <-
#'     "ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.gtf.gz"
#' ref <-
#'     GenomicFeatures::makeTxDbFromGFF(ref)
#' junctions <-
#'     junction_norm(junctions_annot_example)
#' junctions <-
#'     junction_score(junctions)
#' coverage_paths_case <-
#'     list.files("/data/RNA_seq_diag/mito/bw/", full.names = T)[1:2]
#' coverage_paths_control <-
#'     list.files("/data/recount/GTEx_SRP012682/gtex_bigWigs/all_gtex_tissues_raw_bigWigs/",
#'         full.names = T
#'     )[1:2]
#' coverage <- coverage_norm(
#'     junctions,
#'     ref,
#'     unannot_width = 20,
#'     coverage_paths_case,
#'     coverage_paths_control,
#'     coverage_chr_control = "chr"
#' )
#'
#' junctions
#' }
#'
#' @export
coverage_score <- function(junctions, coverage, score_func = .zscore, ...) {

    ##### Check user input is correct #####

    if (!identical(names(coverage), c("case", "control"))) {
        stop("coverage should have the names 'case' and 'control'")
    }

    if (!identical(names(coverage[["case"]]), c("exon_coverage_start", "exon_coverage_end", "intron_coverage")) |
        !identical(names(coverage[["control"]]), c("exon_coverage_start", "exon_coverage_end", "intron_coverage"))) {
        stop("coverage matrices should be named 'exon_coverage_start', 'exon_coverage_end', 'intron_coverage'")
    }

    ##### Score coverage in relation to controls #####

    print(stringr::str_c(Sys.time(), " - Generating coverage abnormality score..."))

    coverage_scores <- .coverage_score(coverage, score_func, ...)

    ##### Obtain regions of greatest coverage dysruption #####

    print(stringr::str_c(Sys.time(), " - Obtaining regions with greatest coverage dysruption..."))

    coverage_region_scores_max <- .coverage_score_max(coverage_scores)

    ##### Store output #####

    colnames(coverage_region_scores_max[["regions"]]) <- dimnames(junctions)[[2]]
    colnames(coverage_region_scores_max[["scores"]]) <- dimnames(junctions)[[2]]
    assays(junctions)[["coverage_region"]] <- coverage_region_scores_max[["regions"]]
    assays(junctions)[["coverage_score"]] <- coverage_region_scores_max[["scores"]]

    print(stringr::str_c(Sys.time(), " - done!"))

    return(junctions)
}

#' Score coverage disruptions
#'
#' \code{.coverage_score}
#'
#' @inheritParams coverage_score
#' @inheritParams junction_score
#'
#' @return
#'
#' @keywords internal
#' @noRd
.coverage_score <- function(coverage, score_func, ...) {
    regions <- c("exon_coverage_start", "exon_coverage_end", "intron_coverage")
    coverage_scores <- vector(mode = "list", length = length(regions))
    names(coverage_scores) <- regions

    for (region in regions) {
        coverage_score_mat <- matrix(
            ncol = ncol(coverage[["case"]][[region]]),
            nrow = nrow(coverage[["case"]][[region]])
        )

        # for every junction/row score coverage
        for (i in seq_along(coverage_score_mat[, 1])) {
            coverage_score_mat[i, ] <-
                score_func(
                    x = coverage[["case"]][[region]][i, ],
                    y = coverage[["control"]][[region]][i, ],
                    ...
                )
        }

        coverage_scores[[region]] <- coverage_score_mat
    }

    names(coverage_scores) <- names(coverage_scores) %>%
        stringr::str_replace("coverage", "coverage_score")

    return(coverage_scores)
}

#' Obtain region of highest coverage disruption for each junction
#'
#' \code{.coverage_score}
#'
#' @inheritParams coverage_score
#' @inheritParams junction_score
#'
#' @return
#'
#' @keywords internal
#' @noRd
.coverage_score_max <- function(coverage_scores) {
    stopifnot(identical(
        names(coverage_scores),
        c(
            "exon_coverage_score_start",
            "exon_coverage_score_end",
            "intron_coverage_score"
        )
    ))

    coverage_scores_max <- matrix(
        ncol = ncol(coverage_scores[[1]]),
        nrow = nrow(coverage_scores[[1]])
    )

    coverage_region_max <- matrix(
        ncol = ncol(coverage_scores[[1]]),
        nrow = nrow(coverage_scores[[1]])
    )

    # loop across samples/cols
    for (i in seq_along(coverage_scores[["exon_coverage_score_start"]][1, ])) {

        ##### Obtain the most dysregulated region #####

        coverage_scores_per_samp <-
            dplyr::tibble(
                exon_coverage_score_start = coverage_scores[["exon_coverage_score_start"]][, i],
                exon_coverage_score_end = coverage_scores[["exon_coverage_score_end"]][, i],
                intron_coverage_score = coverage_scores[["intron_coverage_score"]][, i]
            )

        # find index of the region with highest absolute score
        coverage_region_max[, i] <- coverage_scores_per_samp %>%
            apply(
                MARGIN = 1,
                FUN = function(x) {
                    which.max(abs(x))
                }
            )

        coverage_scores_max[, i] <- coverage_scores_per_samp %>%
            apply(
                MARGIN = 1,
                FUN = function(x) {
                    x[which.max(abs(x))]
                }
            )
    }

    coverage_region_scores_max <- list(
        regions = coverage_region_max,
        scores = coverage_scores_max
    )

    return(coverage_region_scores_max)
}
